letsencrypt/models/letsencrypt.py

```diff
@@ -78,91 +75,123 @@ def validate_domain(self, domain):
+        domain_key_file = self._generate_key(domain)
+        client = self._create_client(account_key)
+        new_reg = NewRegistration(
+            key=account_key.public_key(),
+            terms_of_service_agreed=True)
+        client.new_account(new_reg)
+        csr = self._make_csr(account_key, domain_key_file, domain)
+        authzr = client.new_order(csr)
+        # write the file with the challenge
+        challenge_dir = _get_challenge_dir()
+        if not isdir(challenge_dir):
+            makedirs(challenge_dir)
+        # we don't care what's in the CSR, domains,
+        # whatever challenges letsencrypt sends back, respond
+        # prefer http though.
+        for authorizations in authzr.authorizations:
+            for challenge in authorizations.body.challenges:
+                # we might be given multiple challenges,
+                # we only need to respond to one.
+                if challenge.chall.typ == TYPE_CHALLENGE_HTTP:
+                    self._respond_challenge_http(challenge, account_key)
+                    break
+                elif challenge.chall.typ == TYPE_CHALLENGE_DNS:
+                    self._respond_challenge_dns(challenge, domain)
+                    break
+                else:
+                    _logger.error(
+                        "Unknown challenge received from letsencrypt.")
+        # let them know we are done and they should check
+        from datetime import datetime, timedelta
+        poll_deadline = datetime.now() + timedelta(minutes=5)
+        order_resource = client.finalize_order(authzr, poll_deadline)
+        with open(join(_get_data_dir(), '%s.crt' % domain), 'w') as crt:
+            crt.write(order_resource.fullchain_pem)
+            _logger.info('SUCCESS: Certificate saved :%s', crt.name)
+            ir_config_parameter = self.env['ir.config_parameter']
+            reload_cmd = ir_config_parameter.get_param(
+                'letsencrypt.reload_command', False)
+            if reload_cmd:
+                self.call_cmdline(['sh', '-c', reload_cmd])
+            else:
+                _logger.warning("No reload command defined.")
+
+    def _create_client(self, account_key):
+        net = client.ClientNetwork(account_key)
+        if self.env.context.get('DRY_RUN'):
+            directory_url = V2_STAGING_DIRECTORY_URL
         else:
-            _logger.info('no command defined for reloading webserver, please '
-                         'do it manually in order to apply new certificate')
+            directory_url = V2_DIRECTORY_URL
+        directory_json = requests.get(directory_url).json()
+        return client.ClientV2(directory_json, net)
+
+    def _make_csr(self, account_key, domain_key_file, domain):
+        parameter_model = self.env['ir.config_parameter']
+        altnames = parameter_model.search(
+            [('key', 'like', 'letsencrypt.altname.%')],
+            order='key')
+        valid_domains = [x.value for x in altnames] + [domain]
```

Comment:
I think what needs to happen here is some more smartness:

- sort the list of domains by their components (aka `labels` in DNS speech)
- if we have an asterisk at position `n`, remove all domains with the same amount of labels (so if we have `*.therp.nl` and `www.therp.nl`, this should collapse to `*.therp.nl`)
- keep domains with differing amounts of labels (so if we add `*.runbot.therp.nl` above, it must be preserved)